Fix MSSQL with returns on inserts and deletes on tables with triggers

KalebMatthews opened the pull request:
This change adds the ability to return values on insert and delete statements on tables that contain triggers. Existing insert test cases should cover the testing of these changes.

kibertoad commented:
Since tests weren't previously failing, they cannot be covering these changes. We need new ones that cover the changed behaviour. I can help with writing them, if needed.

KalebMatthews commented:
This change requires a database that has triggers. What is your preferred method for me to go about creating this?  I don't know what databases you have access to in order to reproduce this.  To help and get started, the repo in issue #4151 has a sql script that generates a database with triggers as a testing scenario.  Maybe you could use that?  Let me know and I'll be glad to proceed.

kibertoad commented:
@KalebMatthews We use dockerized databases in CI (see https://github.com/knex/knex/blob/master/scripts/docker-compose.yml). How MSSQL-specific those triggers are? I think it should be perfectly possible to create triggers on the runtime by using `knex.raw`.

KalebMatthews commented:
@kibertoad I will work on getting a containerized test for this using that method.

kibertoad commented:
@KalebMatthews The simplest way to do that is to use `npm run db:start:no-oracle` command, and then run `npm run test:db:no-oracle` to execute integration tests on all DBs. If you would need to only run your own test and not whole integration test suite, use `describe.only`/`it.only`. If your test will come in a new file, make sure to also add it to `https://github.com/knex/knex/blob/master/test/integration/suite.js`

kibertoad commented on lib/dialects/mssql/query/compiler.js:
```diff
@@ -41,8 +41,9 @@ class QueryCompiler_MSSQL extends QueryCompiler {
     const insertValues = this.single.insert || [];
-    let sql = this.with() + `insert into ${this.tableName} `;
     const { returning } = this.single;
+    let sql = this.with() + `${returning ? `declare @out table (${returning
```
this is quite difficult to read. Could you extract whole new ternary block into a separate variable?

KalebMatthews replied:
Agreed. Simplifying and condensing repeated code for readability and maintainability.

kibertoad commented on lib/dialects/mssql/query/compiler.js:
```diff
@@ -80,6 +81,11 @@ class QueryCompiler_MSSQL extends QueryCompiler {
     }
+
+    if (returning) {
+      sql += `; select ${returning.map(value => `[${value}]`).join(",")} from @out`;
```
Same here, mapping function inside string template is difficult to read.

kibertoad commented on lib/dialects/mssql/query/compiler.js:
```diff
@@ -120,10 +126,11 @@ class QueryCompiler_MSSQL extends QueryCompiler {
       sql:
         withSQL +
-        `delete from ${tableName}` +
+        `${returning ? `declare @out table (${returning.map(value => `[${value}] nvarchar(max) null`)
```
Same as above.

kibertoad commented on lib/dialects/mssql/query/compiler.js:
```diff
@@ -163,22 +168,36 @@ class QueryCompiler_MSSQL extends QueryCompiler {
   }
 
+  _buildReturningDeclare(values) {
+    if (values) {
```
why do additional check if you always check value to be set before using these new functions?

KalebMatthews replied:
If there are no return values I wanted to prevent the use of temp tables. So if no values are being returned then it will not add any unused statements. 

There is an issue here though. Truthy statements do not check length of an array. I will make that fix on both of those statements.

kibertoad commented on test/integration/builder/deletes.js:
```diff
@@ -82,5 +82,86 @@ module.exports = function (knex) {
+        let insertedId;
+        let deletedId;
+
+        async function setup() {
```
it would be more idiomatic for mocha if you extracted setup into `before` block and cleanup into `after` block. Probably will need additional `context` or `describe` block for that.

KalebMatthews replied:
I will not be able to reuse the variables to make maintenance a little bit easier. I would either have to break them into global or just hard code. I am okay doing that just want to make sure first.

Would you like that in a new file or just have two describes in that file?

kibertoad replied:
Yeah, I think both new tests could be moved to a new file.

kibertoad replied:
Describes can have layers, so you could put them into top-level describe.

kibertoad commented on test/integration/builder/deletes.js:
```diff
@@ -82,5 +82,86 @@ module.exports = function (knex) {
+        return true;
+      }
+
+      return await knex.transaction(async function () {
```
I'm always wary of returning such large promises because you might start swallowing errors should it ever resolve early for some reason.
Safer approach would be something along the lines
```
let reachedEnd = false
await transaction // sets reachedEnd in the end
expect(reachedEnd).to.be(true)
```

kibertoad commented on test/integration/builder/inserts.js:
```diff
@@ -1817,5 +1817,81 @@ module.exports = function (knex) {
       });
     });
+
+    it('#4152 Should allow returns with inserts on tables with triggers', async function () {
```
since there is lots of reusable setup here, I wonder if both should be moved to more focused file, e. g. 'triggers.js' and have shared `beforeEach`/`afterEach` statements.

KalebMatthews replied:
Okay. I can make it a new file and add it to the suite. I will then make the reused variables global for reuse for any additional tests we need. Thanks for all your help!

kibertoad replied:
Thank you for putting so much effort into fixing this!

KalebMatthews commented:
Copy/Paste fail. Removed.

KalebMatthews commented:
> Is this trigger mssql-specific? if yes, probably skip should be done already on a describe level.

Yes, this is mssql-specific because of the raw sql commands. Where would I put it on the describe level? First line or in the before?

kibertoad commented:
> 
> 
> > Is this trigger mssql-specific? if yes, probably skip should be done already on a describe level.
> 
> Yes, this is mssql-specific because of the raw sql commands. Where would I put it on the describe level? First line or in the before?

Yeah, first line of describe should do the trick.

KalebMatthews commented:
> > > Is this trigger mssql-specific? if yes, probably skip should be done already on a describe level.
> > 
> > 
> > Yes, this is mssql-specific because of the raw sql commands. Where would I put it on the describe level? First line or in the before?
> 
> Yeah, first line of describe should do the trick.

Had to put it in the before. The this object in the describe function did not have a function skip. The before worked fine though. Running test then pushing.

kibertoad commented:
> 
> 
> > > > Is this trigger mssql-specific? if yes, probably skip should be done already on a describe level.
> > > 
> > > 
> > > Yes, this is mssql-specific because of the raw sql commands. Where would I put it on the describe level? First line or in the before?
> > 
> > 
> > Yeah, first line of describe should do the trick.
> 
> Had to put it in the before. The this object in the describe function did not have a function skip. The before worked fine though. Running test then pushing.

You'd need to replace arrow function `() =>` with `function()` for describe callback argument for `this.skip` to work. But if it works in before, that also works :).

kibertoad commented:
```
reason: ReferenceError: returning is not defined
      at QueryCompiler_MSSQL._buildReturningDeclare (/home/travis/build/knex/knex/lib/dialects/mssql/query/compiler.js:189:23)
      at QueryCompiler_MSSQL.insert (/home/travis/build/knex/knex/lib/dialects/mssql/query/compiler.js:45:37)
      at QueryCompiler_MSSQL.toSQL (/home/travis/build/knex/knex/lib/query/compiler.js:61:29)
      at Builder.toSQL (/home/travis/build/knex/knex/lib/query/builder.js:97:44)
      at /home/travis/build/knex/knex/lib/runner.js:31:36
      at /home/travis/build/knex/knex/lib/runner.js:277:24
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

kibertoad commented on test/integration/builder/triggers.js:
```diff
@@ -0,0 +1,202 @@
+
+        // Reset tables for each test
+        beforeEach(async function () {
+            if (knex.client.driverName !== 'mssql') {
```
Do we still get here even if `this.skip` is executed in `before` block?

kibertoad commented on test/integration/builder/triggers.js:
```diff
@@ -0,0 +1,202 @@
+            });
+
+            it('#4152 Should allow returns with inserts on tables with triggers', async function () {
+                if (knex.client.driverName !== 'mssql') {
```
if `this.skip` is executed in `before`, it shouldn't get here, so it shouldn't be necessary.

kibertoad commented on test/integration/builder/triggers.js:
```diff
@@ -0,0 +1,202 @@
+                `);
+            })
+            it('#4152 Should allow returns with deletes on tables with triggers', async function () {
+                if (knex.client.driverName !== 'mssql') {
```
if `this.skip` is executed in `before`, it shouldn't get here, so it shouldn't be necessary.

kibertoad commented:
@KalebMatthews 
```
reason: TypeError: values.map is not a function
      at QueryCompiler_MSSQL._buildReturningDeclare (/home/travis/build/knex/knex/lib/dialects/mssql/query/compiler.js:189:30)
      at QueryCompiler_MSSQL.insert (/home/travis/build/knex/knex/lib/dialects/mssql/query/compiler.js:45:37)
      at QueryCompiler_MSSQL.toSQL (/home/travis/build/knex/knex/lib/query/compiler.js:61:29)
      at Builder.toSQL (/home/travis/build/knex/knex/lib/query/builder.js:97:44)
      at /home/travis/build/knex/knex/lib/runner.js:31:36
      at /home/travis/build/knex/knex/lib/runner.js:277:24
      at runMicrotasks (<anonymous>)
```

KalebMatthews commented:
@kibertoad This solution breaks when a user tries to return *. I saw it in a debug when running the tests to validate something. Is that a valid use case for a returning object for either an insert or delete?

kibertoad commented:
> 
> 
> @kibertoad This solution breaks when a user tries to return *. I saw it in a debug when running the tests to validate something. Is that a valid use case for a returning object for either an insert or delete?

If it used to work in the past, we can't change it without semver major, so I would prefer to keep that supported.

kibertoad commented on lib/dialects/mssql/query/compiler.js:
```diff
@@ -186,17 +186,31 @@ class QueryCompiler_MSSQL extends QueryCompiler {
+      let selections = "";
+
+      if (Array.isArray(values)) {
+        selections = values.map(value => `[${value}]`).join(",");
```
will this work with multiple values? it is supposed to be `[xyz], [abc]`?

KalebMatthews replied:
From what I saw when I debugged this 2 types of values came in ("<data>" and ["data1","data2",...]). This will work with both of these use cases and is tested in multiple pre-existing test cases.

kibertoad replied:
looooots of tests are failing now, though.

KalebMatthews replied:
It is failing because the test is using a string comparison of the generated SQL to validate its responses. This no long equals the generated sql because of the added functionality.  Would you like me to go into each test that is failing using string comparisons and update them accordingly?

Note there are a large number of tests that would need to be changed and might take some time to finalize. Wanted to get you're okay before doing this since this is touching tests that are not a part of my change directly.

kibertoad replied:
So end result is still the same, just SQL is slightly different?
Yeah, unfortunately I don't think it is possible to get CI to green without changing the tests :(

KalebMatthews replied:
Okay. It took awhile but I got the logic modified to handle returning "*" and modified all existing tests that only test the generated sql string to check for the new format.

KalebMatthews commented:
> what is this import for?

Left over from when I was attempting to solve the issue a different way. Removed.

kibertoad commented:
@elhigu Can you also take a look? After I saw what kind of SQL we generate now I am somewhat worried if this is the right direction overall.

KalebMatthews commented:
> so there is a temporary table created for any mssql operation whatsoever now? is that absolutely necessary? wouldn't that incur a noticeable performance hit?

It is only created when there is a returning value passed in. That is what the `if(values && values.length > 0)` check is for.

There is no other way in MSSQL to grab the inserted/updated/deleted values if a trigger is attached because of possible additional modifications to data inside the trigger.

kibertoad commented:
@KalebMatthews Got it. And yeah, obviously there is no way for us to know ahead of time whether or not there is a trigger for that table. @elhigu What do you think, should that be the default behaviour or something that developer somehow triggers?

kibertoad commented:
After thinking about it for a while, I think my personal preference would be to introduce options object for `returning` operation that would have a single parameter `includeTriggerModifications`. If set to true, new logic is used, if not or it is set to false, then it works the old way.

KalebMatthews commented:
As requested I added a 3rd arguments to insert and updates and a second argument for deletes to handle an options argument object. If the object contains true for includeTriggerModifications then the trigger commands will be used instead. 

I have also updated all of the tests to include a trigger variant inside new files and it is re-testing all the use cases with the trigger option set to get a full range of testing.

kibertoad commented:
```
/home/travis/build/knex/knex/lib/dialects/mssql/query/compiler.js:107
    let sql = this.with() + `${this._buildTempTable(returning)}insert into ${this.tableName} `;

        ^
SyntaxError: Identifier 'sql' has already been declared
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/travis/build/knex/knex/lib/dialects/mssql/index.js:12:23)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/travis/build/knex/knex/test/unit/query/builder.js:10:22)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Suite.<anonymous> (/home/travis/build/knex/knex/test/db-less-test-suite.js:16:3)
    at Object.create (/home/travis/build/knex/knex/node_modules/mocha/lib/interfaces/common.js:148:19)
    at context.describe.context.context (/home/travis/build/knex/knex/node_modules/mocha/lib/interfaces/bdd.js:42:27)
    at Object.<anonymous> (/home/travis/build/knex/knex/test/db-less-test-suite.js:13:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/travis/build/knex/knex/test/all-tests-suite.js:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.exports.requireOrImport (/home/travis/build/knex/knex/node_modules/mocha/lib/esm-utils.js:20:12)
    at Object.exports.loadFilesAsync (/home/travis/build/knex/knex/node_modules/mocha/lib/esm-utils.js:33:34)
    at Mocha.loadFilesAsync (/home/travis/build/knex/knex/node_modules/mocha/lib/mocha.js:431:19)
    at singleRun (/home/travis/build/knex/knex/node_modules/mocha/lib/cli/run-helpers.js:125:15)
    at exports.runMocha (/home/travis/build/knex/knex/node_modules/mocha/lib/cli/run-helpers.js:190:10)
    at Object.exports.handler (/home/travis/build/knex/knex/node_modules/mocha/lib/cli/run.js:362:11
```

kibertoad commented:
@KalebMatthews Could you please document new options for the operation in https://github.com/knex/documentation?

KalebMatthews commented:
> @KalebMatthews Could you please document new options for the operations in https://github.com/knex/documentation?

Here ya go: https://github.com/knex/documentation/pull/301

kibertoad commented:
@KalebMatthews Are you aware of performance implications of this particular implementation? I wonder if it's any faster than just SELECTíng affected rows after performing the operation.

KalebMatthews commented:
> @KalebMatthews Are you aware of performance implications of this particular implementation? I wonder if it's any faster than just SELECTíng affected rows after performing the operation.

@kibertoad Selecting the data after you insert is actually two statements...the insert/update plus another select.  On top of that, if you are using primary keys with identity (auto-increment), the OUTPUT statement is the most solid way to make sure you get the primary key values you worked with.  The OUTPUT statement is part of SQL Server's native DML using their specialized "inserted" and "deleted" tables that happen with any DML statement...so you are actually doing the most performant thing I can think of by hooking into the native lifecycle of a MSSQL transaction.  As far as using the declare table statement...I have to rely on the #table to maintain the dynamic nature of the returning * functionality.  So I'm hooking into the native DML lifecycle to get the most correct answer back from the engine.

This definitely has a small hit performance based on what is already there, but this is the most optimal solution for this use case. Putting it behind an options flag is a good solution to protect the current performance but give people the flexibility to get around trigger related issues. This solution is better than a selection because it is as performant if not better and is the only way (without a lot of magic) to pull back the correct returns with identities. Especially if other fields are included in the return.

I ran this by our lead developer who has been working with MSSQL for over 20 years and he can not think of a better solution either. I'm open to suggestions if you have any but I can't imagine running a select afterward would be any more performant or accurate.

kibertoad commented:
@KalebMatthews I'm mostly concerned about the performance hit from creating/dropping temp table (`#out`). Is it negligible, compared to doing a separate SELECT statement without manipulating any tables?

KalebMatthews commented:
> @KalebMatthews I'm mostly concerned about the performance hit from creating/dropping temp table (`#out`). Is it negligible, compared to doing a separate SELECT statement without manipulating any tables?

@kibertoad Yes, it is negligible given this use case. 

Also note we cannot think of a better way to get the correct return when you have a identity field as your primary key with potentially a secondary composite key and you have multiple returnings. Which is also part of our scenario.

kibertoad commented:
@KalebMatthews That is fair. Then this can be merged after documentation is finalized :). Thanks!

KalebMatthews commented:
> @KalebMatthews That is fair. Then this can be merged after documentation is finalized :). Thanks!

Working on that now. Thanks a ton for your help with all of this!

kibertoad commented:
@KalebMatthews How difficult would it be to support `http://knexjs.org/#Builder-returning` as well?

kibertoad commented on test/integration/builder/trigger-updates.js:
```diff
@@ -130,6 +130,47 @@ module.exports = function (knex) {
                 expect(reachedEnd).to.be.true;
             });
+
+            it('#4152 Should allow returns with inserts on tables with triggers using returning function', async function () {
```
should be `updates`, right?

kibertoad commented:
Aaaand it's done.
@KalebMatthews How urgently do you need this released?

KalebMatthews commented:
It's a blocker for a deadline. Whenever y'all are willing would be fine. 

Thanks again for all your help and patience with this. You rock!

kibertoad commented:
@KalebMatthews deadlines are important. Released in 0.21.14

No, y o u  rock! Thank you for making knex better for everyone!

KrisLau commented:
Merges seem to be having the same issue as this: https://github.com/knex/knex/pull/6050
